### PR TITLE
Curl_socket(): fix socktype and protocol

### PR DESCRIPTION
**Pull request by Cering**
for fix https://github.com/curl/curl/issues/9274

### CODE REVIEW

File: lib/connect.c [hunk 1]
```diff
@@ -1607,6 +1607,12 @@ CURLcode Curl_socket(struct Curl_easy *data,
   addr->socktype = (conn->transport == TRNSPRT_TCP) ? SOCK_STREAM : SOCK_DGRAM;
   addr->protocol = conn->transport != TRNSPRT_TCP ? IPPROTO_UDP :
     ai->ai_protocol;
+  if(addr->socktype == SOCK_STREAM && addr->protocol == IPPROTO_UDP)
+   /* If ai from an dns cache whose ai_protocol is IPPROTO_UDP, and want to
+    * create TRNSPRT_TCP socket here, change addr->protocol to IPPROTO_TCP.
+    * Otherwise socket() will return a 'Protocol not supported' error.
+    */
+    addr->protocol = IPPROTO_TCP;
```
Comment (bagder) [inline lib/connect.c:1615]: Can't we simplify this and instead just do this:
~~~diff
--- a/lib/connect.c
+++ b/lib/connect.c
@@ -1603,12 +1603,11 @@ CURLcode Curl_socket(struct Curl_easy *data,
    * if this has been set, before that, it is initialized from parameters.
    */
 
   addr->family = ai->ai_family;
   addr->socktype = (conn->transport == TRNSPRT_TCP) ? SOCK_STREAM : SOCK_DGRAM;
-  addr->protocol = conn->transport != TRNSPRT_TCP ? IPPROTO_UDP :
-    ai->ai_protocol;
+  addr->protocol = conn->transport == TRNSPRT_TCP ? IPPROTO_TCP : IPPROTO_UDP;
   addr->addrlen = ai->ai_addrlen;
 
   if(addr->addrlen > sizeof(struct Curl_sockaddr_storage))
      addr->addrlen = sizeof(struct Curl_sockaddr_storage);
   memcpy(&addr->sa_addr, ai->ai_addr, addr->addrlen);

Comment (Cering) [reply lib/connect.c:1615]: Yes! This is graceful

**Review by bagder — DISMISSED** [review]

**Comment by bagder** [comment]
Argh. I forgot unix domain sockets, and when I tried to adapt to those I spotted they they were wrongly marked TCP so it wasn't that easy, and when setting them to use `TRNSPRT_UNIX` they broke `Curl_socket()` ...

IOW: I did #9276, which actually also is a fix for #9274 but different than this. If you're fine with that, we can close this PR again.

**Comment by Cering** [comment]
will fixed in https://github.com/curl/curl/pull/9276

**Pull request closed without merging**